### PR TITLE
Add horizontal / vertical grid lines filled with colors 

### PR DESCRIPTION
**Pull request by costagolub**
As a feature tagged there [https://github.com/recharts/recharts/issues/860](https://github.com/recharts/recharts/issues/860)

### CODE REVIEW

File: src/cartesian/CartesianGrid.js [hunk 1]
```diff
@@ -174,7 +254,9 @@ class CartesianGrid extends Component {
 
     return (
       <g className="recharts-cartesian-grid">
-        {this.renderBackground()}
```
Comment (xile611) [inline src/cartesian/CartesianGrid.js]: @costagolub We can use `renderBackground` to render a background rectangle. If you remove this function, we can only achieve this by rendering several rectangles which will bring some strange gaps.

File: src/cartesian/CartesianGrid.js [hunk 2]
```diff
@@ -131,6 +136,81 @@ class CartesianGrid extends Component {
...
+  /**
+   * Draw vertical grid stripes filled by colors
+   * @param {Array} verticalPoints either passed in as props or generated from function
+   * @return {Group} Vertical stripes
+   */
+  renderVerticalStripes(verticalPoints) {
+    const { verticalFill } = this.props;
+    if (!verticalFill || !verticalFill.length) { return null; }
+
+    const { fillOpacityVertical, x, y, width, height } = this.props;
```
Comment (xile611) [inline src/cartesian/CartesianGrid.js]: The props `fillOpacityVertical` seems to be very strange.

File: src/cartesian/CartesianGrid.js [hunk 3]
```diff
@@ -131,6 +136,81 @@ class CartesianGrid extends Component {
...
+    let verticalPointsUpdated = verticalPoints;
+    if (x === 0) {
```
Comment (xile611) [inline src/cartesian/CartesianGrid.js]: I can't understand this line. Why not unshift x when x !== verticalPoints[0]?

File: src/cartesian/CartesianGrid.js [hunk 4]
```diff
@@ -131,6 +136,81 @@ class CartesianGrid extends Component {
...
+      verticalPointsUpdated.unshift(0);
+    }
+
+    const items = verticalPointsUpdated.map((entry, i) => {
+      let lineWidth = Math.max(0, (i === 0) ? verticalPointsUpdated[i+1]-x : verticalPointsUpdated[i+1] !== undefined ? verticalPointsUpdated[i+1] - entry : width - entry);
```
Comment (xile611) [inline src/cartesian/CartesianGrid.js]: You need lint for this line.
And you should simplify this line.
```js
const lineWidth = !_.isNil(verticalPointsUpdated[i + 1] ? verticalPointsUpdated[i + 1] - entry :  (x + width - entry);

if (lineWidth <= 0) { return null; } 
```

File: src/cartesian/CartesianGrid.js (hunk 2 above)
Comment (costagolub) [reply src/cartesian/CartesianGrid.js]: What do mean by strange? You think is it better to use the default FillOpacity property both for horizontal and vertical lines?

File: src/cartesian/CartesianGrid.js (hunk 3 above)
Comment (costagolub) [reply src/cartesian/CartesianGrid.js]: Ok, I need to check your suggestion how it works. Thanks!

File: src/cartesian/CartesianGrid.js (hunk 1 above)
Comment (costagolub) [reply src/cartesian/CartesianGrid.js]: @xile611 yes, I actually returned it back in the amended commit.

File: src/cartesian/CartesianGrid.js (hunk 4 above)
Comment (costagolub) [reply src/cartesian/CartesianGrid.js]: ok, I will do, thanks.

File: src/cartesian/CartesianGrid.js (hunk 2 above)
Comment (xile611) [reply src/cartesian/CartesianGrid.js]: @costagolub How about use the default fillOpacity?

**Comment by costagolub** [comment]
#895

**Comment by norman0612** [comment]
will this go out with the next release?

File: src/cartesian/CartesianGrid.js [hunk 5]
```diff
@@ -131,6 +136,81 @@ class CartesianGrid extends Component {
...
+   * @return {Group} Vertical stripes
+   */
+  renderVerticalStripes(verticalPoints) {
+    const { verticalFill } = this.props;
+    if (!verticalFill || !verticalFill.length) { return null; }
+
+    const { fillOpacity, x, y, width, height } = this.props;
+    const verticalPointsUpdated = verticalPoints;
+    if (x !== verticalPoints[0]) {
+      verticalPointsUpdated.unshift(0);
```
Comment (xile611) [inline src/cartesian/CartesianGrid.js]: @costagolub You should unshift `x`.

File: src/cartesian/CartesianGrid.js [hunk 6]
```diff
@@ -131,6 +136,81 @@ class CartesianGrid extends Component {
...
+   * @return {Group} Horizontal stripes
+   */
+  renderHorizontalStripes(horizontalPoints) {
+    const { horizontalFill } = this.props;
+    if (!horizontalFill || !horizontalFill.length) { return null; }
+
+    const { fillOpacity, x, y, width, height } = this.props;
+    const horizontalPointsUpdated = horizontalPoints;
+    if (y !== horizontalPoints[0]) {
+      horizontalPointsUpdated.unshift(0);
```
Comment (xile611) [inline src/cartesian/CartesianGrid.js]: @costagolub It's same here. You should unshift `x`.

**Comment by xile611** [comment]
@costagolub How about create another PR for Tooltip?
It's OK for the feature`horizontal / vertical grid lines filled with colors`. But I'm not sure about the `Tooltip` feature now.

**Comment by xile611** [comment]
@costagolub We has a component `DefaultTooltipContent` now, may be we can add some  TooltipContent to support tooltip has `corner`.

**Comment by costagolub** [comment]
hey @xile611, sure it's not a problem to shift the tooltip component to another PR. 
But do you propose to extend the `defaultTooltipContent `component with the corner?

File: src/cartesian/CartesianGrid.js (hunk 5 above)
Comment (costagolub) [reply src/cartesian/CartesianGrid.js]: @xile611 will fix it, thanks

**Comment by xile611** [comment]
@costagolub By now, I prefer to create another `TooltipContent`.

**Pull request closed without merging**